### ISSUE TITLE
"GradleProperties has not been loaded yet" error when building project with ProjectBuilder

### ISSUE THREAD

**Issue by bingranl** [gradle/gradle#13122]
### Expected Behavior
Gradle introduced new pipe line `providerFactory.gradleProperty` for accessing gradle properties to support instant execution. This new pipeline should be compatible with building Project using ProjectBuilder in unit test. In other words, Gradle should invoke the gradle property loading mechanism at `projectBuilder.build()`   

### Current Behavior
When applying android gradle plugin to the project and accessing some gradle property with the new API, we get "GradleProperties has not been loaded yet" error.

	at org.gradle.initialization.DefaultGradlePropertiesController$NotLoaded.gradleProperties(DefaultGradlePropertiesController.java:74)
	at org.gradle.initialization.DefaultGradlePropertiesController$SharedGradleProperties.gradleProperties(DefaultGradlePropertiesController.java:59)
	at org.gradle.initialization.DefaultGradlePropertiesController$SharedGradleProperties.find(DefaultGradlePropertiesController.java:50)
	at org.gradle.api.internal.provider.sources.GradlePropertyValueSource.obtain(GradlePropertyValueSource.java:36)
	at org.gradle.api.internal.provider.sources.GradlePropertyValueSource.obtain(GradlePropertyValueSource.java:24)
	at org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$ValueSourceProvider.obtainValueFromSource(DefaultValueSourceProviderFactory.java:248)
	at org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$ValueSourceProvider.lambda$calculateOwnValue$0(DefaultValueSourceProviderFactory.java:225)
	at org.gradle.internal.Try.ofFailable(Try.java:39)
	at org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$ValueSourceProvider.calculateOwnValue(DefaultValueSourceProviderFactory.java:225)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.getOrNull(AbstractMinimalProvider.java:103)
	at com.android.build.gradle.options.ProjectOptions.getValue(ProjectOptions.java:144)
	at com.android.build.gradle.internal.DependencyConfigurator.configureGeneralTransforms(DependencyConfigurator.kt:103)
	at com.android.build.gradle.internal.plugins.BasePlugin.createAndroidTasks(BasePlugin.java:601)
	at com.android.build.gradle.internal.plugins.PluginExtensionsKt.runAfterEvaluate(PluginExtensions.kt:24)
	at com.android.build.gradle.internal.plugins.PluginExtensionsKt.runAfterEvaluate$default(PluginExtensions.kt:23)
	at com.android.build.gradle.internal.tests.ConfigurationExtensionTest.setUp(ConfigurationExtensionTest.kt:235)

### Context
We are migrating how we obtain gradle properties from `project.getExtensions().getExtraProperties()` to `providerFactory.gradleProperty`. This issue blocks us as our unit tests no longer work.

**Comment by bmuskalla** [comment]
@eskatos Is there something I can help with getting this resolved? I'm hitting this as some of our own integration tests start to fail as one of the services for `JavaBasePlugin` starts to look at environment variables. Can unblock myself with a workaround for now.

**Comment by eskatos** [comment]
@bmuskalla, just to clarify, are you saying an "integration test", or did you observe this on a test based on `ProjectBuilder` as reported in this issue?

**Comment by bmuskalla** [comment]
Oops, I meant unit test. Specifically, a new test in `JavaBasePluginTest`

**Comment by petedoyle** [comment]
Anyone else running Dependency Verification? I've recently had the error pop up, possibly after updating from 6.3 to 6.5.1 (not 100% sure on this). Deleting gradle/verification-metadata.xml immediately fixes it for me.

**Comment by bendb-instacart** [comment]
I just hit this in a unit test using `ProjectBuilder`.  The test was attempting to apply the Android Gradle Plugin; I had just bumped the AGP version from 4.1.0-beta03 to 4.2.0-alpha12.  It's hard for me to say what changed, because nothing in my code did change - just the dependency being applied.

**Comment by benjamin-bader** [comment]
I just hit this in a unit test using ProjectBuilder. The test was attempting to apply the Android Gradle Plugin; I had just bumped the AGP version from 4.1.0-beta03 to 4.2.0-alpha12. It's hard for me to say what changed, because nothing in my code did change - just the dependency being applied.

**Comment by benjamin-bader** [comment]
To make it a little more concrete, here's a link to a line of code in a PR, and a link to the actual test failure:

code: https://github.com/KeepSafe/dexcount-gradle-plugin/blob/agp-4.2.0/src/test/groovy/com/getkeepsafe/dexcount/DexCountExtensionSpec.groovy#L69
failing test: https://github.com/KeepSafe/dexcount-gradle-plugin/pull/357/checks?check_run_id=1178328113#step:5:93

note that in the PR, nothing about the test changes - the only code changes here are some slight re-arrangments due to API changes in AGP.  Particularly, nothing remotely related to this test changes, aside from the version bump.

**Comment by benjamin-bader** [comment]
This is pretty gross, but a workaround seems to be wrapping all my ProjectBuilder test setup within `project.gradle.settingsEvaluated`, like so:

```groovy
def "my spec"() {
  given:
  project.gradle.settingsEvaluated {
    apply plugin: "com.android.application"
    apply plugin: "com.bendb.plugin"
    // etc
  }

  when:
  project.evaluate() // etc
}
```

Naturally this is _far_ from ideal, but it unblock things for now.

**Comment by eskatos** [comment]
Here's a workaround until this gets fixed:
```java
Project project = ProjectBuilder.builder()......build()
(project as ProjectInternal).services.get(GradlePropertiesController.class).loadGradlePropertiesFrom(projectDir)
// ....
```
It relies on internal APIs though.

**Comment by bric3** [comment]
Same issue on https://github.com/gradle/native-platform, I cannot even import the project in IJ to get the dependencies.
@eskatos Where should we put this ?

Is it because I'm not part of gradle (the projects uses `com.gradle.enterprise.gradle-enterprise-conventions-plugin`) ?

**Comment by eskatos** [comment]
@bric3, I'm not sure how this related, could you please open an issue on `gradle/native-platform` so that the relevant team can take a look at that build? The use of the Gradle enterprise on this open source project should not prevent you from using it.

**Comment by bric3** [comment]
@eskatos Done !
Thanks for the feedback.

**Comment by oehme** [comment]
Since several people posted issues related to dependency verification: This issue is only about the `ProjectBuilder` API.

The issue with dependency verification in init scripts will be fixed by https://github.com/gradle/gradle/pull/16757

**Comment by eskatos** [comment]
Fixed in https://github.com/gradle/gradle/pull/16757
This will be in Gradle 7.1

**Comment by bingranl** [comment]
Thanks for working on this issue! I tested with Gradle 7.1 and it doesn't seem to work for me.

1. In our unit test set up, we add a "gradle.properties" file with some properties before we invoke `projectBuilder.build()`
2. After invoking `projectBuilder.build()`, I try to access the property by doing `project.getProviders().gradleProperty("name").get()`, but couldn't get the property(error message: no value for that provider)

Did I miss something?

**Comment by oehme** [comment]
There is no support for gradle.properties files in ProjectBuilder tests. Just set the property on the Project object directly.

**Comment by bingranl** [comment]
Thanks for your reply, Stefan! Are you suggesting setting the property via `project.setProperty(name, value)`? If I do that, I am able to find the property from `project.getProperties`, but not from `project.getProviders().gradleProperty("name").get()`. Or you are talking about a different API?

**Comment by oehme** [comment]
Ah sorry I had missed that part in your original request. I had only fixed the exception that @eskatos had linked by binding an empty GradleProperties instance for ProjectBuilder tests. That of course isn't enough for your use case. Either we need to support gradle.properties files or provide a way to pass properties programatically so they will be available via the Provider API.

@eskatos which option would you prefer?

**Comment by eskatos** [comment]
`ProjectBuilder` unit tests should be writable without fiddling with the filesystem most of the time so I'd say that having an API on `ProjectBuilder` to define available properties is tempting.

**Comment by bingranl** [comment]
Thanks for working on it, folks! Do we want to re-open this issue or create a new one to keep track of it?

**Comment by oehme** [comment]
I think we should just reopen it, since your original request was not fixed. @eskatos can you do that please, since I don't have the permissions.

**Comment by eskatos** [comment]
I would create another issue instead because this one is in the list of things fixed for a past Gradle version already. Maybe the description would need to be adjusted to make it clear what was done in 7.1 though.

**Comment by bingranl** [comment]
Hey folks, I have created a new issue https://github.com/gradle/gradle/issues/17638. Please take a look! Thanks!
